pajbot/managers/twitter.py

```diff
@@ -162,25 +201,105 @@ def check_twitter_connection(self):
-        if self.twitter_client:
-            try:
-                public_tweets = self.twitter_client.user_timeline(username)
-                for tweet in public_tweets:
-                    if not tweet.text.startswith("RT ") and tweet.in_reply_to_screen_name is None:
-                        # Tweepy returns naive datetime object (but it's always UTC)
-                        # .replace() makes it timezone-aware :)
-                        created_at = tweet.created_at.replace(tzinfo=datetime.timezone.utc)
-                        tweet_message = stringify_tweet(tweet)
-                        return f"{tweet_message} ({time_since(utils.now().timestamp(), created_at.timestamp(), time_format='short')} ago)"
-            except Exception:
-                log.exception("Exception caught while getting last tweet")
-                return "FeelsBadMan"
-        else:
-            return "Twitter not set up FeelsBadMan"
-
-        return "FeelsBadMan"
-
     def quit(self):
         if self.twitter_stream:
             self.twitter_stream.disconnect()
+
+
+# PBTwitterManager reads live tweets from a pajbot tweet-provider (https://github.com/pajbot/tweet-provider) instead of Twitter's streaming API
+class PBTwitterManager(GenericTwitterManager):
+    relevant_users = []
+    bot = None
+    client = None
+    tweepy = None
+
+    def __init__(self, bot):
+        super().__init__(bot)
+
+        PBTwitterManager.bot = bot
+        PBTwitterManager.tweepy = self.twitter_client
+
+        self.listener = PBTwitterManager
+
+        if "twitter" not in bot.config:
+            return
+
+        self.reload()
+
+        log.info("pajbot twitter manager initialized")
+
+        from twisted.internet import reactor
+
+        class ClientProtocol(WebSocketClientProtocol):
+            def onOpen(self):
+                PBTwitterManager.client = self
+
+                log.info("Connected to tweet-provider")
+
+                user_ids = [PBTwitterManager.tweepy.get_user(screen_name=e).id for e in PBTwitterManager.relevant_users]
+
+                msg = {"type": "set_subscriptions", "data": user_ids}
+
+                self.sendMessage(json.dumps(msg).encode("utf8"))
+
+            def onMessage(self, payload, isBinary):
+                if isBinary:
+                    return
+
+                message = json.loads(payload)
+                if message["type"] == "tweet":
+                    tweet = message["data"]
+                    if (
+                        tweet["user"]["screen_name"].lower() in PBTwitterManager.relevant_users
+                        and not tweet["text"].startswith("RT ")
+                        and tweet["in_reply_to_screen_name"] is None
+                    ):
+                        tweet_message = tweet_provider_stringify_tweet(tweet)
+                        PBTwitterManager.bot.say(
+                            f"B) New cool tweet from {tweet['user']['screen_name']}: {tweet_message}"
+                        )
```

Comment:
Honestly this should be in a bot.execute_now call but I can also live without it